### PR TITLE
Support building against macOS universal binaries

### PR DESCRIPTION
**Pull request by clowder**
FIXES: #1087

A "universal binary" is a fat binary which contains multiple builds for different architectures. When Postgres is packaged in this format it's necessary to slice out the build for the current architecture before trying to parse it. The program terminates early if the universal binary doesn't support the current architecture.

The list of supported architectures in `slice_arch32` is limited by [`object`][object-arch], which supports fewer than [Rust][rust-arch].

The universal binary test fixture was simply chosen because of it's small size relative to other binaries in Postgres.

[^1]: https://github.com/gimli-rs/object/issues/454#issuecomment-1200668989

[rust-arch]: http://doc.rust-lang.org/1.68.2/std/env/consts/constant.ARCH.html
[object-arch]: https://docs.rs/object/latest/object/read/macho/trait.FatArch.html#method.architecture

### CODE REVIEW

**Comment by eeeebbbbrrrr** [comment]
Hey, this is cool.  Thanks a bunch.  I'm sure @thomcc will be excited to review this in detail.

> The universal binary test fixture was simply chosen because of it's small size relative to other binaries in Postgres.

Any chance we can rename this to anything other than `pg_config`?  I just know that name will cause confusion down the road since pgrx uses the real pg_config everywhere.

**Comment by clowder** [comment]
@eeeebbbbrrrr Thanks for the quick reply!

> Any chance we can rename this to anything other than `pg_config`? I just know that name will cause confusion down the road since pgrx uses the real pg_config everywhere.

Happy to do this. Just to clarify, it's literally a universal binary version of the pg_config utility. With that in mind, would reshuffling the parts be enough? For example `cargo-pgrx/tests/fixtures/pg_config.macos-universal`. Or should we just give it some opaque name? For example `cargo-pgrx/tests/fixtures/macos-universal`.

**Comment by eeeebbbbrrrr** [comment]
Yeah, i'd prefer it be called something like `test-macos-universal-binary` or similar.  I suppose we don't care what the binary actually is.

Also, I can't tell in the PR, but please make sure it doesn't accidentally have the execute bit (+x) set.

**Comment by clowder** [comment]
> Yeah, i'd prefer it be called something like `test-macos-universal-binary` or similar. I suppose we don't care what the binary actually is.

Updated. I opted for `macos-universal-binary` because `test` was already in the path.

> Also, I can't tell in the PR, but please make sure it doesn't accidentally have the execute bit (+x) set.

It did. This has been fixed now.

**Comment by eeeebbbbrrrr** [comment]
Nice, thanks!  I just realized I never approved you for CI, so now that's happening.  I'd still like @thomcc to give this a look and approve.  He's more experienced with this kind of thing than I am.

**Comment by clowder** [comment]
> Yeah, glad I turned on CI. It doesn't like something here: https://github.com/tcdi/pgrx/actions/runs/5269612633/jobs/9528132717#step:10:160

Sorry about that, I missed something renaming that fixture. Sorted now.

File: cargo-pgrx/src/command/schema.rs [hunk 1]
```diff
@@ -556,3 +558,82 @@ fn create_stub(
...
+                    panic!("Failed to slice architecture '{arch}' from universal binary.")
+                }
+            }
+        }
+        _ => object::File::parse(&*data),
+    }
+}
+
+fn slice_arch32<'a>(data: &'a [u8], arch: &str) -> Option<&'a [u8]> {
+    let target = match arch {
```
Comment (thomcc) [inline cargo-pgrx/src/command/schema.rs:581]: Most of these almost certainly don't work, but I suppose it's harmless.

**Review by thomcc — APPROVED** [review]
Sorry for the delay, this looks fine to me.

**Pull request merged**